src/lib.rs

```diff
@@ -343,59 +345,68 @@ pub enum CargoOpt {
     SomeFeatures(Vec<String>),
 }
 
-/// Obtain metadata only about the root package and don't fetch dependencies
-///
-/// # Parameters
-///
-/// - `manifest_path`: Path to the manifest.
-pub fn metadata(manifest_path: Option<impl AsRef<Path>>) -> Result<Metadata> {
-    metadata_run(manifest_path, false, None)
+/// A builder for configurating `cargo metadata` invocation.
+#[derive(Debug, Clone, Default)]
+pub struct MetadataCommand {
+    manifest_path: Option<PathBuf>,
+    current_dir: Option<PathBuf>,
+    no_deps: bool,
+    features: Option<CargoOpt>,
 }
 
-/// Obtain metadata only about the root package and dependencies
-///
-/// # Parameters
-///
-/// - `manifest_path`: Path to the manifest.
-/// - `deps`: Whether to include dependencies.
-pub fn metadata_deps(manifest_path: Option<impl AsRef<Path>>, deps: bool) -> Result<Metadata> {
-    metadata_run(manifest_path, deps, None)
-}
-
-/// The main entry point to obtaining metadata
-///
-/// # Parameters
-///
-/// - `manifest_path`: Path to the manifest.
-/// - `deps`: Whether to include dependencies.
-/// - `feat`: Which features to include, `None` for defaults.
-pub fn metadata_run(manifest_path: Option<impl AsRef<Path>>, deps: bool, features: Option<CargoOpt>) -> Result<Metadata> {
-    let manifest_path = manifest_path.as_ref().map(|p| p.as_ref());
-    let cargo = env::var("CARGO").unwrap_or_else(|_| String::from("cargo"));
-    let mut cmd = Command::new(cargo);
-    cmd.arg("metadata");
-
-    if !deps {
-        cmd.arg("--no-deps");
+impl MetadataCommand {
+    /// Creates a default `cargo metadata` command, which will look for
+    /// `Cargo.toml` in the ancestors of the current directory.
+    pub fn new() -> MetadataCommand {
+        MetadataCommand::default()
     }
-
-    if let Some(features) = features {
-        match features {
-            CargoOpt::AllFeatures => cmd.arg("--all-features"),
-            CargoOpt::NoDefaultFeatures => cmd.arg("--no-default-features"),
-            CargoOpt::SomeFeatures(ftrs) => cmd.arg(format!("--fatures {:?}", ftrs)),
-        };
+    /// Path to `Cargo.toml`
+    pub fn manifest_path(&mut self, path: impl AsRef<Path>) -> &mut MetadataCommand {
+        self.manifest_path = Some(path.as_ref().to_path_buf());
+        self
     }
-
-    cmd.args(&["--format-version", "1"]);
-    if let Some(manifest_path) = manifest_path {
-        cmd.arg("--manifest-path").arg(manifest_path.as_os_str());
+    /// Current directory of the `cargo metadata` process.
+    pub fn current_dir(&mut self, path: impl AsRef<Path>) -> &mut MetadataCommand {
+        self.current_dir = Some(path.as_ref().to_path_buf());
+        self
+    }
+    /// Output information only about the root package and don't fetch dependencies.
+    pub fn no_deps(&mut self, no_deps: bool) -> &mut MetadataCommand {
```

Comment:
That's true in general, but in this specific case I think it makes sense to invert:

* this matches the semantics of the underling cargo sub command, which has a `no_deps` flag
* the common case is that you do want the dependencies, ~so it would be poor ergonomics to force almost everyone to call `.with_deps(true)`~. The last one is wrong: we can just make `true` a default for `with_deps`

Given this, should I flip to `with_deps`?

Reply:
Alternatively you could remove the `bool` argument so one would just call `builder.no_deps()` which would do the expected thing and not emit anything about dependencies. That would mirror the cargo flag (it's not like you need to pass `--no-deps=false` or `--no-deps=true`)